File: benchmark/benchmark-rng.py
import timeit
import numpy as np

setupNP = """
import numpy as np
"""

setupSP = """
from scipy.stats import norm, rayleigh, rice
"""

def run_test(title, test, setup):
    print '{:>40}'.format(title+':'),
    x = timeit.Timer(test, setup=setup).repeat(5,1)
    print '{:>9.3f} {:>9.3f} {:>9.3f}'.format(min(x), np.mean(x), max(x))

sizes = [
            {'title': 'Small 32x32x100 sample (flat)', 'n': 32*32*100},
            {'title': 'Small 32x32x100 sample (shaped)', 'n': '(32,32,100)'},
            {'title': 'Full 64x64x39x259 sample (flat)', 'n': 64*64*39*259},
        ]

for size in sizes:
    print '\n{:<40}'.format(size['title'])
    print '{:<41}{:>9} {:>9} {:>9}'.format('', 'min','mean','max')
    run_test('NumPy normal dist',
             "xnp = np.random.normal(scale=1.2, size={})".format(size['n']), setup=setupNP)
    run_test('NumPy rayleigh dist',
             "xnp = np.random.rayleigh(scale=1.2, size={})".format(size['n']), setup=setupNP)
    run_test('SciPy normal dist',
             "xscipy = norm.rvs(scale=1.2, size={})".format(size['n']), setup=setupSP)
    run_test('SciPy rayleigh dist',
             "xscipy = rayleigh.rvs(scale=1.2, size={})".format(size['n']), setup=setupSP)
    run_test('SciPy rice dist',
             "xscipy = rice.rvs(scale=1.2, b=1, size={})".format(size['n']), setup=setupSP)

File: benchmark/benchmark-rng.R
library(neuRosim)
library(VGAM)

run_test <- function(title, test) {
    cat(sprintf('%30s:',title))
    x <- replicate(5,system.time(eval(test)))
    cat(sprintf('%9.3f %9.3f %9.3f',min(x[3,]),mean(x[3,]),max(x[3,])),'\n')

}

sizes = list(
             list(title='Small 32x32x100 sample', n=32*32*100),
             list(title='Full 64*64*39*259 sample', n=64*64*39*259)
             )

for (i in 1:length(sizes)) {
    cat('\n', sizes[[i]][['title']], '\n')
    cat(sprintf('%30s ',''))
    cat(sprintf('%9s %9s %9s','min','mean','max'),'\n')
    run_test('Builtin normal dist',
             expression(y <- rnorm(n=sizes[[i]][['n']], sd=1.2)))
    run_test('VGAM rayleigh dist',
             expression(y <- rrayleigh(n=sizes[[i]][['n']], scale=1.2)))
    run_test('VGAM rice dist',
             expression(y <- VGAM::rrice(n=sizes[[i]][['n']], vee=1, sigma=1.2)))
    run_test('neuRosim rice dist',
             expression(y <- neuRosim::rrice(n=sizes[[i]][['n']], vee=1, sigma=1.2)))
}
